logstash/lib/logstash/outputs/loki.rb

```diff
@@ -0,0 +1,262 @@
+
+  ## 'Extra labels to add to all log streams'
+  config :external_labels, :valide => :hash,  :default => {}, :required => false
+
+
+  public
+  def register
+    @uri = URI.parse(@url + '/loki/api/v1/push')
+    unless @uri.is_a?(URI::HTTP) || @uri.is_a?(URI::HTTPS)
+      raise LogStash::ConfigurationError, "url parameter must be valid HTTP, currently '#{@url}'"
+    end
+
+    @logger.info("New Loki output", :class => self.class.name)
+
+    @Channel = Concurrent::Channel
+    @batches = {}
+    @entries = @Channel.new
+    @exclude_labels += ["message", "@timestamp"] 
+    @retry_http = 0
+    @delay = 1
+    if ssl_cert?
+      load_ssl
+      validate_ssl_key
+    end
+
+    @Channel.go{run()}
+  end
+
+  def ssl_cert?
+    !@key.nil? && !@cert.nil?
+  end
+
+  def load_ssl
+    @cert = OpenSSL::X509::Certificate.new(File.read(@cert)) if @cert
+    @key = OpenSSL::PKey.read(File.read(@key)) if @key
+  end
+
+  def validate_ssl_key
+    if !@key.is_a?(OpenSSL::PKey::RSA) && !@key.is_a?(OpenSSL::PKey::DSA)
+      raise "Unsupported private key type #{key.class}"
+    end
+  end
+
+  def ssl_opts(uri)
+    opts = {
+      use_ssl: uri.scheme == 'https'
+    }
+
+    if !@cert.nil? && !@key.nil?
+      opts = opts.merge(
+        verify_mode: OpenSSL::SSL::VERIFY_PEER,
+        cert: @cert,
+        key: @key
+      )
+    end
+
+    unless @ca_cert.nil?
+      opts = opts.merge(
+        ca_file: @ca_cert
+      )
+    end
+    opts
+  end 
+
+  def run()
+	  min_wait_checkfrequency = 1/1000 #1 millisecond
+	  max_wait_checkfrequency = @batch_wait / 10
+	  if max_wait_checkfrequency < min_wait_checkfrequency
+		  max_wait_checkfrequency = min_wait_checkfrequency
+    end
+
+    max_wait_check = Concurrent::Channel.tick(max_wait_checkfrequency)
+    loop do
+      Concurrent::Channel.select do |s|
+        s.take(@entries) { |e|
+          if !@batches.has_key?(e.tenant_id)
+            @batches[e.tenant_id] = Batch.new(e)
+            next
+          end
+
+          batch = @batches[e.tenant_id]
+          line = e.entry['line']
+          if batch.size_bytes_after(line) > @batch_size
+            send(e.tenant_id, batch)
+            @batches[e.tenant_id] = Batch.new(e)
+            next
+          end
+          batch.add(e)
+        }
+        s.take(max_wait_check) {
+          # Send all batches whose max wait time has been reached
+          @batches.each { |tenant_id, batch|
+            if batch.age() < @batch_wait
+              next
+            end
+            send(tenant_id, batch)
+            @batches.delete(tenant_id)
+          }
+        }
+      end
+    end 
+  end
+
+  public
+  def receive(event)
+    labels = {}
+    event_hash = event.to_hash
+    lbls = extract_labels(event_hash, labels, "")
+    data_labels = lbls.merge(@external_labels)
+
+    entry_hash = {
+      "ts" => event.get("@timestamp").to_i * (10**9) +  Time.new.nsec,
+      "line" => event.get("message").to_s
+    }
+
+    @Channel.go do
```

Comment:
I don’t think you need a go here ?

Reply:
Not required. Will remove

Reply:
Actually `Channel.go` is required here. Since we are using unbuffered channel(i.e without specifying capacity), if we don't use `Channel.go` it hangs there without proceeding with the code

Reply:
Hanging is fine, if it is hanging it is probably waiting for the previous batch to be sent.

If you keep it this way, it means for each even, (million of them) create a go routine. That can be huge for the system.

You may want to use buffered but I think it's better to wait, this also signal, don't send more we're busy to the plugin.

Reply:
Yeah, got your point. Will make the change